include/aspect/material_model/rheology/elasticity.h

```diff
@@ -89,9 +89,9 @@ namespace aspect
            * fill an additional material model outputs objects with the elastic force terms.
            */
           void
-          fill_elastic_force_outputs (const MaterialModel::MaterialModelInputs<dim> &in,
-                                      const std::vector<double> &average_elastic_shear_moduli,
-                                      MaterialModel::MaterialModelOutputs<dim> &out) const;
+          fill_elastic_outputs (const MaterialModel::MaterialModelInputs<dim> &in,
```

Comment:
Thanks a lot for the review :-) I have thought over the question whether cell averaging affects the elastic force, and now my answer is yes. To calculate the elastic force, we use MaterialModelOutputs::viscosities as effective viscosity. However, in both of the two material models (Viscoelastic and ViscoPlastic) that include elasticity, function fill_elastic_force_outputs() is called inside evaluate(), which means that the viscosity will be changed by MaterialAveraging::average() afterwards. This will increase numerical error when the viscosity varies dramatically in space. Therefore, I think it would be safer to avoid using MaterialModelOutputs::viscosities in function evaluate(). In the case of elasticity, we may just store the elastic viscosity $\eta^{ve} = G\Delta t^e$ in ElasticOutputs, and calculate the elastic force by $F^e_{ij} = -(\eta^{eff}\tau_{ij}^0) / \eta^{ve}$ when assembling the system RHS. What do you think about this modification?

Reply:
Hi Yimin, hi all,
1. At the moment, we do not allow MaterialAvaraging (cell averaging of material output) except for the viscosity, because averaging of the RHS term is not implemented: https://github.com/geodynamics/aspect/blob/d04495c3439e1cfa003350214414b52ee2cc7ef9/source/material_model/rheology/elasticity.cc#L209C1-L224C55.
2. As you noted, the RHS elastic force term is computed using the un-cell-averaged viscosity, while the Stokes equation is further assembled with the averaged viscosity (if it switched on). (https://github.com/geodynamics/aspect/blob/d04495c3439e1cfa003350214414b52ee2cc7ef9/source/simulator/assembly.cc#L565C1-L569C80)
That leaves me with several thoughts:
1. Do you think averaging the elastic force term would remedy the discrepency between averaged viscosity and a force term using the unaveraged viscosity?
2. If instead we do as you say and store the elastic viscosity, should this viscosity also be averaged for consistency?
3. I think it would be nice to leave the computation of the force term in the material model, especially as it will become slightly more complicated with #4370, and maybe with other formulations in the future.
4. At the moment, the Stokes assembly for the Picard solver is not adapted, but MaterialAveraging of the viscosity is allowed. This would mean even for the Picard solver, there is a potential inconsistency that should be remedied.

Reply:
We've just discussed the MaterialAveraging further, and since at the moment any material averaging is inconsistent with RHS terms that do not used averaged properties, I'll open a PR that calls a function with the averaged properties to fill the RHS *Outputs instead of filling them in evaluate(). This function will be called after evaluate() in the assembly, which means it can use the averaged properties, e.g. viscosity, to compute the RHS term.

Reply:
Hi @anne-glerum ,

Thanks for considering this issue with so much attention. Now this PR is completed and the Newton method has reached its expected performance (not the overall efficiency, since there are many details that can be improved). According to my experiments, it is _necessary_ to average the viscosity used for evaluating the elastic force when the viscosity is averaged in the system matrix. About your questions,

> At the moment, we do not allow MaterialAvaraging (cell averaging of material output) except for the viscosity, because averaging of the RHS term is not implemented

Only the viscosity has an impact on the LHS, so it does not matter if the other properties are averaged.

> Do you think averaging the elastic force term would remedy the discrepency between averaged viscosity and a force term using the unaveraged viscosity?

I have not tested that, but I think it is better to average the viscosity alone, since the deviatoric stress to be multiplied to the viscosity is different at each quadrature point.

> If instead we do as you say and store the elastic viscosity, should this viscosity also be averaged for consistency?

It is a good question. I have not considered it, since I have not run a model with nonuniform elastic shear modulus. Theoretically, I think it is not necessary to average the elastic viscosity. The two quantities requiring elastic viscosity are
$$F^e_{ij} = \frac{\eta^{\text{eff}}}{\eta^e}\tau_{ij}^0,\quad\varepsilon_{ij}^{ve}=\varepsilon_{ij}+\frac{\tau_{ij}^0}{2\eta^e},$$
which are on the RHS and LHS, respectively. The final viscosity $\eta^{\text{eff}}$ must be averaged in $F^e_{ij}$, because it is averaged on the LHS. But $\eta^e$ exists on both LHS and RHS, so it should be OK if it is averaged or not averaged simultaneously on both sides.

> I think it would be nice to leave the computation of the force term in the material model

I think so too. In this PR I do the averaging inside functions Elasticity::fill_elastic_outputs() and Elasticity::fill_reaction_terms(). However, with such an implementation we will meet a problem: for "project to Q1" average we should know the quadrature formula, but inside evaluate() we have no information of it (for instance, in function compute_artificial_viscosity() the quadrature formula becomes QTrepezoid). Currently I simply excluded the case "project to Q1" and added a "TODO" mark.

> I'll open a PR that calls a function with the averaged properties to fill the RHS *Outputs instead of filling them in evaluate(). This function will be called after evaluate() in the assembly, which means it can use the averaged properties

I have tried this idea, but also encountered with a problem: should we fill the RHS outputs in evaluate() or not? If we do, then there is duplicate computation, which is ugly; if we do not, then there are too many cases to take care of. For example, if we use particle-in-cell method, then the reaction terms are evaluated at each single point in Particle::Property::ElasticStress::initialize_one_particle_property() without averaging, and it looks absurd to call evaluate() and then call another function to fill the reaction terms. That is the reason why I choose to average the viscosity inside evaluate() currently.

Reply:
> I think so too. In this PR I do the averaging inside functions Elasticity::fill_elastic_outputs() and Elasticity::fill_reaction_terms(). However, with such an implementation we will meet a problem: for "project to Q1" average we should know the quadrature formula, but inside evaluate() we have no information of it (for instance, in function compute_artificial_viscosity() the quadrature formula becomes QTrepezoid). Currently I simply excluded the case "project to Q1" and added a "TODO" mark.

Therefore it would be great to find a solution that works for all averaging options _and_ not just for the ElasticOutputs force term, but for every RHS term that can be added.

> I have tried this idea, but also encountered with a problem: should we fill the RHS outputs in evaluate() or not? If we do, then there is duplicate computation, which is ugly; if we do not, then there are too many cases to take care of. For example, if we use particle-in-cell method, then the reaction terms are evaluated at each single point in Particle::Property::ElasticStress::initialize_one_particle_property() without averaging, and it looks absurd to call evaluate() and then call another function to fill the reaction terms. That is the reason why I choose to average the viscosity inside evaluate() currently.

I envision it as follows in the Stokes assembly:

1. Call evaluate(). Do not fill the ElasticOutputs (or other RHS term Outputs) in evaluate().
2. MaterialAverage the MaterialModelOutputs
3. Call a function that fills the ElasticOutputs with the averaged MaterialModelOutputs

The reaction terms and reaction rates can still be filled during evaluate, they are separate from the RHS Stokes terms.

Reply:
Okay, after more discussion with @YiminJin  and Juliane, we decided that it is the responsibility of the material model itself to correctly MaterialAverage properties that are used to compute other material model output. So in the viscoplastic material model, the viscosity can be MaterialAveraged, so if it is used to compute a force or reaction term, it should first be MaterialAveraged.